pkg/k8s/shared_resources.go

```diff
@@ -0,0 +1,90 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package k8s
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/fields"
+
+	"github.com/cilium/cilium/pkg/hive"
+	"github.com/cilium/cilium/pkg/hive/cell"
+	cilium_api_v2alpha1 "github.com/cilium/cilium/pkg/k8s/apis/cilium.io/v2alpha1"
+	"github.com/cilium/cilium/pkg/k8s/client"
+	"github.com/cilium/cilium/pkg/k8s/resource"
+	slim_corev1 "github.com/cilium/cilium/pkg/k8s/slim/k8s/api/core/v1"
+	"github.com/cilium/cilium/pkg/k8s/utils"
+	nodeTypes "github.com/cilium/cilium/pkg/node/types"
+	"github.com/cilium/cilium/pkg/option"
+)
+
+var (
+	// SharedResourceCell provides a set of shared handles to Kubernetes resources used throughout the
+	// Cilium agent. Each of the resources share a client-go informer and backing store so we only
+	// have one watch API call for each resource kind and that we maintain only one copy of each object.
+	//
+	// If you need a filtered view to a resource (e.g. objects in certain namespace) and it already
+	// has a more general resource, then use stream.Filter() to filter the event stream rather than
+	// creating a new resource.
+	//
+	// See pkg/k8s/resource/resource.go for documentation on the Resource[T] type.
+	SharedResourcesCell = cell.Module(
+		"k8s-shared-resources",
+		"Shared Kubernetes resources",
+
+		cell.Provide(
+			serviceResource,
+			localNodeResource,
+			namespaceResource,
+			lbIPPoolsResource,
+		),
+	)
+)
+
+type SharedResources struct {
+	cell.In
+	LocalNode  resource.Resource[*corev1.Node]
+	Services   resource.Resource[*slim_corev1.Service]
+	Namespaces resource.Resource[*slim_corev1.Namespace]
+	LBIPPools  resource.Resource[*cilium_api_v2alpha1.CiliumLoadBalancerIPPool]
+}
+
+func serviceResource(lc hive.Lifecycle, cs client.Clientset) (resource.Resource[*slim_corev1.Service], error) {
+	if !cs.IsEnabled() {
+		return nil, nil
+	}
+	optsModifier, err := utils.GetServiceListOptionsModifier(option.Config)
+	if err != nil {
+		return nil, err
+	}
+	lw := utils.ListerWatcherFromTyped[*slim_corev1.ServiceList](cs.Slim().CoreV1().Services(""))
+	lw = utils.ListerWatcherWithModifier(lw, optsModifier)
+	return resource.New[*slim_corev1.Service](lc, lw), nil
+}
+
+func localNodeResource(lc hive.Lifecycle, cs client.Clientset) (resource.Resource[*corev1.Node], error) {
+	if !cs.IsEnabled() {
+		return nil, nil
+	}
+	lw := utils.ListerWatcherFromTyped[*corev1.NodeList](cs.CoreV1().Nodes())
+	lw = utils.ListerWatcherWithFields(lw, fields.ParseSelectorOrDie("metadata.name="+nodeTypes.GetName()))
```

Comment:
The whole local node think doesn't really match the 

> //If you need a filtered view to a resource (e.g. objects in certain namespace) and it already
> // has a more general resource, then use stream.Filter() to filter the event stream rather than
> // creating a new resource.

comment does it? Is there no place where we need info about all of the nodes like in IPCache or something like that?

Reply:
Ah yeah. Plus these don't expose an Observable either. Killing the comment.
No we don't need all `v1.Node`, but we do need all `CiliumNode`s.